cocos/core/game.ts

```diff
@@ -610,214 +463,160 @@ export class Game extends EventTarget {
-                const bitNum = log2(layer.value);
-                Layers.addLayer(layer.name, bitNum);
-            }
-        }
-
-        return this._initEngine().then(() => {
-            if (!EDITOR || legacyCC.GAME_VIEW) {
+        // DONT change the order unless you know what's you doing
+        return Promise.resolve()
+            .then(() => {
+                this.emit(Game.EVENT_PRE_BASE_INIT);
+                return this.onPreBaseInitDelegate.dispatch();
+            })
+            .then(() => {
+                const debugMode = config.debugMode || debug.DebugMode.NONE;
+                debug._resetDebugSetting(debugMode);
+                sys.init();
                 this._initEvents();
-            }
-
-            if (legacyCC.director.root && legacyCC.director.root.dataPoolManager) {
-                legacyCC.director.root.dataPoolManager.jointTexturePool.registerCustomTextureLayouts(config.customJointTextureLayouts);
-            }
-            return this._engineInited;
-        });
-    }
-
-    /**
-     * @en Run game with configuration object and onStart function.
-     * @zh 运行游戏，并且指定引擎配置和 onStart 的回调。
-     * @param onStart - function to be executed after game initialized
-     */
-    public run(onStart?: Game.OnStart): Promise<void>;
-
-    public run (configOrCallback?: Game.OnStart | IGameConfig, onStart?: Game.OnStart) {
-        // To compatible with older version,
-        // we allow the `run(config, onstart?)` form. But it's deprecated.
-        let initPromise: Promise<boolean> | undefined;
-        if (typeof configOrCallback !== 'function' && configOrCallback) {
-            initPromise = this.init(configOrCallback);
-            this.onStart = onStart ?? null;
-        } else {
-            this.onStart = configOrCallback ?? null;
-        }
-        garbageCollectionManager.init();
-
-        return Promise.resolve(initPromise)
-            .then(() => Promise.resolve(builtinResMgr.initBuiltinRes(this._gfxDevice!)))
-            .then(() => this._setRenderPipelineNShowSplash()).then(() => {
-                if (!HTML5) {
-                    // @ts-expect-error access private method.
-                    screen._init({
-                        configOrientation: 'auto',
-                        exactFitScreen: true,
+            })
+            .then(() => {
+                this.emit(Game.EVENT_POST_BASE_INIT);
+                return this.onPostBaseInitDelegate.dispatch();
+            })
+            // settings initialization
+            .then(() => {
+                this.emit(Game.EVENT_PRE_SETTINGS_INIT);
+                return this.onPreSettingsInitDelegate.dispatch();
+            })
+            .then(() => settings.init(config.settingsPath, config.overrideSettings))
+            .then(() => {
+                this.emit(Game.EVENT_POST_SETTINGS_INIT);
+                return this.onPostSettingsInitDelegate.dispatch(settings);
+            })
+            // Infrastructure region
```

Comment:
This directive comment would do collapse? Did you mean `// #region Infrastructure`

Reply:
Yeah, it's better